File: binary_search.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

class Data {
    long key;
    String value;

    Data(long key, String value) {
        this.key = key;
        this.value = value;
    }
}

public class binary_search {

    public static int binarySearch(List<Data> list, long targetKey) {
        int left = 0, right = list.size() - 1;
        while (left <= right) {
            int mid = (left + right) >>> 1;  // Avoid overflow
            long midKey = list.get(mid).key;
            if (midKey == targetKey) {
                return mid;
            } else if (midKey < targetKey) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return -1;
    }

    public static List<Data> readCSV(String filename) throws IOException {
        List<Data> dataList = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);
                if (parts.length == 2) {
                    long key = Long.parseLong(parts[0].trim());
                    String value = parts[1].trim();
                    dataList.add(new Data(key, value));
                }
            }
        }
        return dataList;
    }

    public static void writeSearchTimes(String filename, double bestTime, double avgTime, double worstTime) throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            writer.write(String.format("Best case running time: %.6e seconds%n", bestTime));
            writer.write(String.format("Average case running time: %.6e seconds%n", avgTime));
            writer.write(String.format("Worst case running time: %.6e seconds%n", worstTime));
        }
    }

    public static void main(String[] args) {
        String inputFile = "merge_sort_1000000.csv";
        String outputFile = "binary_search_1000000.txt";

        try {
            List<Data> data = readCSV(inputFile);
            int datasetSize = data.size();
            if (datasetSize == 0) {
                System.err.println("Dataset is empty!");
                return;
            }

            int n = 100000;  // Number of searches to perform (set as desired)

            // Prepare targets for each case
            long bestTarget = data.get(datasetSize / 2).key;
            long avgTarget = data.get(Math.min(datasetSize / 2 + 1, datasetSize - 1)).key;
            long worstTarget = -1L;  // Non-existent element

            // Best case timing
            long start = System.nanoTime();
            for (int i = 0; i < n; i++) {
                binarySearch(data, bestTarget);
            }
            long end = System.nanoTime();
            double bestTime = (end - start) / 1e9 / n;  // average time in seconds

            // Average case timing
            start = System.nanoTime();
            for (int i = 0; i < n; i++) {
                binarySearch(data, avgTarget);
            }
            end = System.nanoTime();
            double avgTime = (end - start) / 1e9 / n;

            // Worst case timing
            start = System.nanoTime();
            for (int i = 0; i < n; i++) {
                binarySearch(data, worstTarget);
            }
            end = System.nanoTime();
            double worstTime = (end - start) / 1e9 / n;

            // Write results to output file
            writeSearchTimes(outputFile, bestTime, avgTime, worstTime);

            // Print the message to the command prompt
            System.out.println("Results stored in binary_search_1000000.txt");

        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: binary_search.py
import csv
import time

class Data:
    def __init__(self, key, value):
        self.key = int(key)
        self.value = value

def binary_search(arr, target_key):
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid].key == target_key:
            return mid
        elif arr[mid].key < target_key:
            left = mid + 1
        else:
            right = mid - 1
    return -1

def read_csv(filename):
    data_list = []
    with open(filename, newline='') as csvfile:
        reader = csv.reader(csvfile)
        for row in reader:
            if len(row) == 2 and row[0].isdigit():
                data_list.append(Data(row[0].strip(), row[1].strip()))
    return data_list

def write_search_times(filename, best_time, avg_time, worst_time):
    with open(filename, 'w') as f:
        f.write(f"Best case running time: {best_time:.6e} seconds\n")
        f.write(f"Average case running time: {avg_time:.6e} seconds\n")
        f.write(f"Worst case running time: {worst_time:.6e} seconds\n")

def main():
    input_file = "merge_sort_1000000.csv"
    output_file = "binary_search_1000000.txt"

    data = read_csv(input_file)
    dataset_size = len(data)
    if dataset_size == 0:
        print("Dataset is empty!")
        return

    # Explicit number of searches to perform
    n = 100000  # For example, 100,000 searches; you can change this as needed

    # Prepare targets safely
    middle_index = dataset_size // 2
    best_target = data[middle_index].key
    avg_target = data[min(middle_index + 1, dataset_size - 1)].key
    worst_target = -1  # Not in dataset

    # Run n searches for best case
    start = time.perf_counter()
    for _ in range(n):
        binary_search(data, best_target)
    end = time.perf_counter()
    best_time = (end - start) / n

    # Run n searches for average case
    start = time.perf_counter()
    for _ in range(n):
        binary_search(data, avg_target)
    end = time.perf_counter()
    avg_time = (end - start) / n

    # Run n searches for worst case
    start = time.perf_counter()
    for _ in range(n):
        binary_search(data, worst_target)
    end = time.perf_counter()
    worst_time = (end - start) / n

    write_search_times(output_file, best_time, avg_time, worst_time)
    print(f"Performed {n} searches per case.")
    print(f"Running times saved to {output_file}")

if __name__ == "__main__":
    main()
